README.md
# MUonESim_v2
# MUonESim_v3
# MUonESim_v19


History
10/12/2020 Version 1: Project created.

10/1/2021 Version 2: Energy > 1 GeV cut included for all histograms. Process ID is saved in Histogram 11, where: 	          
processID = 1; //ionization
processID = 2; //pair production
processID = 3; //Bremsstrahlung
processID = 4; //nuclear interaction

19/1/2021 Version 3: Axis range in some energy histograms increased.

26/1/2021 Version 4: 2D plot axes swapped. nTuple added.

26/1/2021 Version 5: Angles and energies calculated for all charged particles at target's exit.

27/1/2021 Version 6: Charged track multiplicity counted.

2/2/2021 Version 7: Angles and energies calculated for each process separately.

17/2/2021 Version 8: Ntuple removed.

25/3/2021 Version 14: Angle between e+ and e- from pair production added. 2D plot with energies of e+ and e-.

30/3/2021 Version 15: Test version: Trying to activate separately muNucl.

31/3/2021 Version 16: Physics list from TestEm17 added instead of the old one.

1/7/2021 Version 17: Optional changes: Graphite --> Carbon, detector thickness: 1.5 cm --> 1.3 cm

12/7/2021 Version 18: Muon energy changed from 150 GeV to 160 GeV.

26/7/2021 Version 19: Target's size changed to 10 cm X 10 cm X 1.3 cm. New physics list: FTFP_BERT_EMZ.


src/EventAction.cc
//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
// $Id: EventAction.cc 82401 2014-06-18 14:43:54Z gcosmo $
//
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

#include "EventAction.hh"

#include "DetectorConstruction.hh"
#include "RunAction.hh"
#include "HistoManager.hh"

#include "G4Event.hh"
#include "G4Run.hh"
#include "G4RunManager.hh"
#include "G4EmCalculator.hh"
#include "G4Step.hh"

#include "G4Material.hh"
#include "G4NistManager.hh"
#include "G4Box.hh"
#include "G4LogicalVolume.hh"
#include "G4PVPlacement.hh"
#include "G4Region.hh"
#include "G4ProductionCuts.hh"

#include "G4GeometryManager.hh"
#include "G4PhysicalVolumeStore.hh"
#include "G4LogicalVolumeStore.hh"
#include "G4SolidStore.hh"

#include "G4UnitsTable.hh"
#include "G4SystemOfUnits.hh"
#include "G4VisAttributes.hh"
#include "G4UImanager.hh"

#include "Randomize.hh"
#include <iomanip>

#include <stdio.h>
#include <stdlib.h>

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

EventAction::EventAction(DetectorConstruction* DA, RunAction* RA)
:G4UserEventAction(), fDetectorconstruction(DA), fRunAction(RA)
{}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

EventAction::~EventAction()
{}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void EventAction::BeginOfEventAction(const G4Event* evt)
{

 fEnergyExitTargetPrimary = fEnergyExitTargetSecondaryPositron = fEnergyExitTargetSecondaryElectron = fEnergyExitTargetSecondaryMuonPlus = fEnergyExitTargetSecondaryMuonMinus = fEnergyExitTargetCharged = fEnergyExitTargetChargedCut05GeV = fEnergyExitTargetChargedCut1GeV = fEnergyExitTargetIoni = fEnergyExitTargetPP1 = fEnergyExitTargetPP2 = fEnergyExitTargetIoniCut05GeV = fEnergyExitTargetPP1Cut05GeV = fEnergyExitTargetPP2Cut05GeV = fEnergyExitTargetIoniCut1GeV = fEnergyExitTargetPP1Cut1GeV = fEnergyExitTargetPP2Cut1GeV = 0.;
 fAnglePrimary = fAngleSecondaryPositron = fAngleSecondaryElectron = fAngleSecondaryMuonPlus = fAngleSecondaryMuonMinus = fAngleCharged = fAngleChargedCut05GeV = fAngleChargedCut1GeV = fAngleIoni = fAnglePP1 = fAnglePP2 = fAngleIoniCut05GeV = fAnglePP1Cut05GeV = fAnglePP2Cut05GeV = fAngleIoniCut1GeV = fAnglePP1Cut1GeV = fAnglePP2Cut1GeV = 0.;
fAnglee = 0.;

 fBoolPrim = fBoolSecPositron = fBoolSecElectron = fBoolSecMuonPlus = fBoolSecMuonMinus = fBoolCharged = fBoolChargedCut05GeV = fBoolChargedCut1GeV = fBooleIoni = fBoolePP1 = fBoolePP2 = fBooleIoniCut05GeV = fBoolePP1Cut05GeV = fBoolePP2Cut05GeV = fBooleIoniCut1GeV = fBoolePP1Cut1GeV = fBoolePP2Cut1GeV = 0;
 fNbTracks = fNbTracksCut05GeV = fNbTracksCut1GeV = 0;
 
 fePlusx = fePlusy = fePlusz = feMinusx = feMinusy = feMinusz = 0.;

}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void EventAction::EndOfEventAction(const G4Event* evt)
{
 // get event ID
 //G4int evtNb = evt->GetEventID();
 //G4cout 
    //<< "\n Event ID = " 
    //<< evtNb
    //<< G4endl;

 /*fRunAction->AddREnergyExitTargetPrimary(fEnergyExitTargetPrimary);
 fRunAction->AddREnergyExitTargetSecondaryPositron(fEnergyExitTargetSecondaryPositron);
 fRunAction->AddREnergyExitTargetSecondaryElectron(fEnergyExitTargetSecondaryElectron);
 fRunAction->AddREnergyExitTargetMuonPlus(fEnergyExitTargetMuonPlus);
 fRunAction->AddREnergyExitTargetMuonMinus(fEnergyExitTargetMuonMinus);

 fRunAction->AddRAnglePrimary(fAnglePrimary);
 fRunAction->AddRAngleSecondaryPositron(fAngleSecondaryPositron);
 fRunAction->AddRAngleSecondaryElectron(fAngleSecondaryElectron);
 fRunAction->AddRAngleMuonPlus(fAngleMuonPlus);
 fRunAction->AddRAngleMuonMinus(fAngleMuonMinus);*/

 G4AnalysisManager* analysisManager = G4AnalysisManager::Instance();
 analysisManager->SetVerboseLevel(1);
 
 if (fBoolPrim == 1)  {
    analysisManager->FillH1(1, fEnergyExitTargetPrimary);
    analysisManager->FillH1(6, fAnglePrimary);
    /*analysisManager->FillNtupleDColumn(1, fEnergyExitTargetPrimary);
    analysisManager->FillNtupleDColumn(6, fAnglePrimary);*/
 }
 if (fBoolSecPositron == 1)  {
    analysisManager->FillH1(2, fEnergyExitTargetSecondaryPositron);
    analysisManager->FillH1(7, fAngleSecondaryPositron);
    /*analysisManager->FillNtupleDColumn(2, fEnergyExitTargetSecondaryPositron);
    analysisManager->FillNtupleDColumn(7, fAngleSecondaryPositron);*/
 }
 if (fBoolSecElectron == 1)  {
    analysisManager->FillH1(3, fEnergyExitTargetSecondaryElectron);
    analysisManager->FillH1(8, fAngleSecondaryElectron);
    /*analysisManager->FillNtupleDColumn(3, fEnergyExitTargetSecondaryElectron);
    analysisManager->FillNtupleDColumn(8, fAngleSecondaryElectron);*/
 }
 if ((fBoolSecPositron == 1) && (fBoolSecElectron == 1))  {
    fAnglee = acos(fePlusx*feMinusx + fePlusy*feMinusy + fePlusz*feMinusz);
    analysisManager->FillH1(33, fAnglee);
    
    /*G4cout 
    	<< "\n fePlusx = " 
    	<< fePlusx
    	<< "\n fePlusy = " 
    	<< fePlusy
    	<< "\n fePlusz = " 
    	<< fePlusz
    	<< "\n feMinusx = " 
    	<< feMinusx
    	<< "\n feMinusy = " 
    	<< feMinusy
    	<< "\n feMinusz = " 
    	<< feMinusz
    	<< "\n fAnglee = " 
    	<< fAnglee
    	<< G4endl;*/
 }
 if (fBoolSecMuonPlus == 1)  {
    analysisManager->FillH1(4, fEnergyExitTargetSecondaryMuonPlus);
    analysisManager->FillH1(9, fAngleSecondaryMuonPlus);
    /*analysisManager->FillNtupleDColumn(4, fEnergyExitTargetSecondaryMuonPlus);
    analysisManager->FillNtupleDColumn(9, fAngleSecondaryMuonPlus);*/
 }
 if (fBoolSecMuonMinus == 1)  {
    analysisManager->FillH1(5, fEnergyExitTargetSecondaryMuonMinus);
    analysisManager->FillH1(10, fAngleSecondaryMuonMinus);
    /*analysisManager->FillNtupleDColumn(5, fEnergyExitTargetSecondaryMuonMinus);
    analysisManager->FillNtupleDColumn(10, fAngleSecondaryMuonMinus);*/
 }
 if (fBoolCharged == 1)  {
    analysisManager->FillH1(12, fEnergyExitTargetCharged);
    analysisManager->FillH1(13, fAngleCharged);
    /*analysisManager->FillNtupleDColumn(12, fEnergyExitTargetCharged);
    analysisManager->FillNtupleDColumn(13, fAngleCharged);*/
    
    analysisManager->FillH1(18, fNbTracks);
    //analysisManager->FillNtupleDColumn(18, fNbTracks);
 }
 if (fBoolChargedCut05GeV == 1)  {
    analysisManager->FillH1(14, fEnergyExitTargetChargedCut05GeV);
    analysisManager->FillH1(15, fAngleChargedCut05GeV);
    /*analysisManager->FillNtupleDColumn(14, fEnergyExitTargetChargedCut05GeV);
    analysisManager->FillNtupleDColumn(15, fAngleChargedCut05GeV);*/
    
    analysisManager->FillH1(19, fNbTracksCut05GeV);
    //analysisManager->FillNtupleDColumn(19, fNbTracksCut05GeV);
 }
 if (fBoolChargedCut1GeV == 1)  {
    analysisManager->FillH1(16, fEnergyExitTargetChargedCut1GeV);
    analysisManager->FillH1(17, fAngleChargedCut1GeV);
    /*analysisManager->FillNtupleDColumn(16, fEnergyExitTargetChargedCut1GeV);
    analysisManager->FillNtupleDColumn(17, fAngleChargedCut1GeV);*/
    
    analysisManager->FillH1(20, fNbTracksCut1GeV);
    //analysisManager->FillNtupleDColumn(20, fNbTracksCut1GeV);
 }
 
 if (fBooleIoni == 1)  {
    analysisManager->FillH1(21, fEnergyExitTargetIoni);
    analysisManager->FillH1(27, fAngleIoni);
    
    /*analysisManager->FillNtupleDColumn(21, fEnergyExitTargetIoni);
    analysisManager->FillNtupleDColumn(27, fAngleIoni);*/
 }
 if (fBoolePP1 == 1)  {
    analysisManager->FillH1(22, fEnergyExitTargetPP1);
    analysisManager->FillH1(28, fAnglePP1);
    
    /*analysisManager->FillNtupleDColumn(22, fEnergyExitTargetPP1);
    analysisManager->FillNtupleDColumn(28, fAnglePP1);*/
 }
 if (fBoolePP2 == 1)  {
    analysisManager->FillH1(22, fEnergyExitTargetPP2);
    analysisManager->FillH1(28, fAnglePP2);
    
    /*analysisManager->FillNtupleDColumn(22, fEnergyExitTargetPP2);
    analysisManager->FillNtupleDColumn(28, fAnglePP2);*/
 }
 
 if (fBooleIoniCut05GeV == 1)  {
    analysisManager->FillH1(23, fEnergyExitTargetIoniCut05GeV);
    analysisManager->FillH1(29, fAngleIoniCut05GeV);
    
    /*analysisManager->FillNtupleDColumn(23, fEnergyExitTargetIoniCut05GeV);
    analysisManager->FillNtupleDColumn(29, fAngleIoniCut05GeV);*/
 }
 if (fBoolePP1Cut05GeV == 1)  {
    analysisManager->FillH1(24, fEnergyExitTargetPP1Cut05GeV);
    analysisManager->FillH1(30, fAnglePP1Cut05GeV);
    
    /*analysisManager->FillNtupleDColumn(24, fEnergyExitTargetPP1Cut05GeV);
    analysisManager->FillNtupleDColumn(30, fAnglePP1Cut05GeV);*/
 }
 if (fBoolePP2Cut05GeV == 1)  {
    analysisManager->FillH1(24, fEnergyExitTargetPP2Cut05GeV);
    analysisManager->FillH1(30, fAnglePP2Cut05GeV);
    
    /*analysisManager->FillNtupleDColumn(24, fEnergyExitTargetPP2Cut05GeV);
    analysisManager->FillNtupleDColumn(30, fAnglePP2Cut05GeV);*/
 }
 
 if (fBooleIoniCut1GeV == 1)  {
    analysisManager->FillH1(25, fEnergyExitTargetIoniCut1GeV);
    analysisManager->FillH1(31, fAngleIoniCut1GeV);
    
    /*analysisManager->FillNtupleDColumn(25, fEnergyExitTargetIoniCut1GeV);
    analysisManager->FillNtupleDColumn(31, fAngleIoniCut1GeV);*/
 }
 if (fBoolePP1Cut1GeV == 1)  {
    analysisManager->FillH1(26, fEnergyExitTargetPP1Cut1GeV);
    analysisManager->FillH1(32, fAnglePP1Cut1GeV);
    
    /*analysisManager->FillNtupleDColumn(26, fEnergyExitTargetPP1Cut1GeV);
    analysisManager->FillNtupleDColumn(32, fAnglePP1Cut1GeV);*/
 }
 if (fBoolePP2Cut1GeV == 1)  {
    analysisManager->FillH1(26, fEnergyExitTargetPP2Cut1GeV);
    analysisManager->FillH1(32, fAnglePP2Cut1GeV);
    
    /*analysisManager->FillNtupleDColumn(26, fEnergyExitTargetPP2Cut1GeV);
    analysisManager->FillNtupleDColumn(32, fAnglePP2Cut1GeV);*/
 }

 if ((fBoolPrim == 1) && (fBoolSecPositron == 1))  {
    analysisManager->FillH2(1, fAngleSecondaryPositron, fAnglePrimary);
 }
 if ((fBoolPrim == 1) && (fBoolSecElectron == 1))  {
    analysisManager->FillH2(2, fAngleSecondaryElectron, fAnglePrimary);
    /*G4cout 
       << "\n Filling histogram 2."
       << G4endl;*/
 }
 if ((fBoolPrim == 1) && (fBoolSecMuonPlus == 1))  {
    analysisManager->FillH2(3, fAngleSecondaryMuonPlus, fAnglePrimary);
 }
 if ((fBoolPrim == 1) && (fBoolSecMuonMinus == 1))  {
    analysisManager->FillH2(4, fAngleSecondaryMuonMinus, fAnglePrimary);
 }
 
 if ((fBoolPrim == 1) && (fBooleIoni == 1))  {
    analysisManager->FillH2(5, fAngleIoni, fAnglePrimary);
 }
 if ((fBoolPrim == 1) && (fBoolePP1 == 1))  {
    analysisManager->FillH2(6, fAnglePP1, fAnglePrimary);
 }
 if ((fBoolPrim == 1) && (fBoolePP2 == 1))  {
    analysisManager->FillH2(6, fAnglePP2, fAnglePrimary);
 }
 
 if ((fBoolPrim == 1) && (fBooleIoniCut05GeV == 1))  {
    analysisManager->FillH2(7, fAngleIoniCut05GeV, fAnglePrimary);
 }
 if ((fBoolPrim == 1) && (fBoolePP1Cut05GeV == 1))  {
    analysisManager->FillH2(8, fAnglePP1Cut05GeV, fAnglePrimary);
 }
 if ((fBoolPrim == 1) && (fBoolePP2Cut05GeV == 1))  {
    analysisManager->FillH2(8, fAnglePP2Cut05GeV, fAnglePrimary);
 }
 
 if ((fBoolPrim == 1) && (fBooleIoniCut1GeV == 1))  {
    analysisManager->FillH2(9, fAngleIoniCut1GeV, fAnglePrimary);
 }
 if ((fBoolPrim == 1) && (fBoolePP1Cut1GeV == 1))  {
    analysisManager->FillH2(10, fAnglePP1Cut1GeV, fAnglePrimary);
 }
 if ((fBoolPrim == 1) && (fBoolePP2Cut1GeV == 1))  {
    analysisManager->FillH2(10, fAnglePP2Cut1GeV, fAnglePrimary);
 }
 if ((fBoolPrim == 1) && (fBoolSecPositron == 1) && (fBoolSecElectron == 1))  {
    analysisManager->FillH2(11, fAngleSecondaryElectron, fAngleSecondaryPositron);
 }
 
 
 //analysisManager->AddNtupleRow();

 //Visualize event if there is a condition
 //if (condition)  {
     //G4cout 
         //<< [condition] fullfilled
         //<< G4endl;
     //G4EventManager* evMan = G4EventManager::GetEventManager();
     //evMan->KeepTheCurrentEvent();
 //}

 //G4cout 
    //<< "\n End of event! "
    //<< G4endl;
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......



src/HistoManager.cc
//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
//
// $Id: HistoManager.cc 72242 2013-07-12 08:44:19Z gcosmo $
// 
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

#include <CLHEP/Units/SystemOfUnits.h>
#include <sstream>

#include "HistoManager.hh"
#include "G4UnitsTable.hh"
#include "G4SystemOfUnits.hh"

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

HistoManager::HistoManager()
  :fFileName("MUonESim")
{
  Book();
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

HistoManager::~HistoManager()
{
  delete G4AnalysisManager::Instance();
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void HistoManager::Book()
{
  // Create or get analysis manager
  // The choice of analysis technology is done via selection of a namespace
  // in HistoManager.hh
  // Creating a tree container to handle histograms and ntuples.
  // This tree is associated to an output file.
  //
  G4AnalysisManager* analysisManager = G4AnalysisManager::Instance();
  analysisManager->SetFileName(fFileName);
  analysisManager->SetVerboseLevel(1);
  analysisManager->SetActivation(true);   //enable inactivation of histograms

  
  // Define histograms start values
  // const G4int kNbofStrips = 1016;
  const G4int kMaxHisto = 34;
  const G4int kMaxHisto2 = 12;

  const G4String id[] = { "000", "001", "002", "003", "004", "005", "006", "007", "008", "009", "010", "011", "012", "013", "014", "015", "016", "017", "018", "019", "020", "021", "022", "023", "024", "025", "026", "027", "028", "029", "030", "031", "032", "033" };
  const G4String id2[] = { "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11"};

  const G4String title[] =
                { "dummy",                                                       //0
                  "Energy of primary muons at the exit of the target (simulation)",   //1
                  "Energy of secondary positrons at the exit of the target, E > 1 GeV (simulation)",   //2
                  "Energy of secondary electrons at the exit of the target, E > 1 GeV (simulation)",   //3
                  "Energy of secondary #mu^{+} at the exit of the target, E > 1 GeV (simulation)",   //4
                  "Energy of secondary #mu^{-} at the exit of the target, E > 1 GeV (simulation)",   //5
                  "Angle of primary muons at the exit of the target (simulation, smaller than 30 mrad)",   //6
                  "Angle of secondary positrons at the exit of the target, E > 1 GeV (simulation, smaller than 30 mrad)",   //7
                  "Angle of secondary electrons at the exit of the target, E > 1 GeV (simulation, smaller than 30 mrad)",   //8
                  "Angle of secondary #mu^{+} at the exit of the target, E > 1 GeV (simulation, smaller than 30 mrad)",   //9
                  "Angle of secondary #mu^{-} at the exit of the target, E > 1 GeV (simulation, smaller than 30 mrad)",   //10
                  "Process ID for the creation of secondary particles, E > 1 GeV (simulation)",  //11
                  "Energy of all charged particles at the exit of the target (simulation)",   //12
                  "Angle of all charged particles at the exit of the target (simulation)",   //13
                  "Energy of charged particles at the exit of the target, E > 0.5 GeV (simulation)",   //14
                  "Angle of charged particles at the exit of the target, E > 0.5 GeV (simulation)",   //15
                  "Energy of charged particles at the exit of the target, E > 1 GeV (simulation)",   //16
                  "Angle of charged particles at the exit of the target, E > 1 GeV (simulation)",   //17
                  "Multiplicity of charged tracks with angle < 30 mrad (simulation)",   //18
                  "Multiplicity of charged tracks with angle < 30 mrad, E > 0.5 GeV (simulation)",   //19
                  "Multiplicity of charged tracks with angle < 30 mrad, E > 1 GeV (simulation)",   //20
                  "Energy of secondary e at the exit of the target, ionization (simulation)",   //21   
                  "Energy of secondary e at the exit of the target, pair production (simulation)",   //22
                  "Energy of secondary e at the exit of the target, ionization, E > 0.5 GeV (simulation)",   //23  
                  "Energy of secondary e at the exit of the target, pair production, E > 0.5 GeV (simulation)",   //24 
                  "Energy of secondary e at the exit of the target, ionization, E > 1 GeV (simulation)",   //25 
                  "Energy of secondary e at the exit of the target, pair production, E > 1 GeV (simulation)",   //26
                  "Angle of secondary e at the exit of the target, ionization (simulation)",   //27  
                  "Angle of secondary e at the exit of the target, pair production (simulation)",   //28
                  "Angle of secondary e at the exit of the target, ionization, E > 0.5 GeV (simulation)",   //29 
                  "Angle of secondary e at the exit of the target, pair production, E > 0.5 GeV (simulation)",   //30 
                  "Angle of secondary e at the exit of the target, ionization, E > 1 GeV (simulation)",   //31 
                  "Angle of secondary e at the exit of the target, pair production, E > 1 GeV (simulation)",   //32
                  "Angle between secondary e+ and e- tracks at the exit of the target, E > 1 GeV (simulation)"   //33
                };

  const G4String title2[] =
                { "dummy",                                                     //0
                  "Angle of primary muon vs. angle of secondary positron (simulation)",   //1
                  "Angle of primary muon vs. angle of secondary electron (simulation)",   //2
                  "Angle of primary muon vs. angle of secondary #mu^{+} (simulation)",    //3
                  "Angle of primary muon vs. angle of secondary #mu^{-} (simulation)",    //4
                  "Angle of primary muon vs. angle of secondary e, ionization (simulation)",   //5
                  "Angle of primary muon vs. angle of secondary e, pair production (simulation)",   //6
                  "Angle of primary muon vs. angle of secondary e, ionization, E > 0.5 GeV (simulation)",   //7
                  "Angle of primary muon vs. angle of secondary e, pair production, E > 0.5 GeV (simulation)",   //8
                  "Angle of primary muon vs. angle of secondary e, ionization, E > 1 GeV (simulation)",   //9
                  "Angle of primary muon vs. angle of secondary e, pair production, E > 1 GeV (simulation)",   //10
                  "Energy of secondary e+ vs. energy of secondary e-, E > 1 GeV (simulation)"   //11
                };
            
  // Default values (to be reset via /analysis/h1/set command)               
  G4int nbins = 100;
  G4double vmin = -0.5;
  G4double vmax = 100.5;

  // Create all histograms as inactivated 
  // as we have not yet set nbins, vmin, vmax
  for (G4int k=0; k<kMaxHisto; k++) {
    G4int ih = analysisManager->CreateH1(id[k], title[k], nbins, vmin, vmax);
    analysisManager->SetH1Activation(ih, false);
  }

  for (G4int k2=0; k2<kMaxHisto2; k2++) {
    G4int ih2 = analysisManager->CreateH2(id2[k2], title2[k2], nbins, vmin, vmax, nbins, vmin, vmax);
    analysisManager->SetH2Activation(ih2, false);
  }

  G4int ih0 = analysisManager->CreateH2("0", "dummy", nbins, vmin, vmax,  //real point
						      nbins, vmin, vmax);   //reconstructed point
  analysisManager->SetH2Activation(ih0, false);
  
  // nTuples
  /*analysisManager->SetNtupleDirectoryName("ntuple");
  analysisManager->SetFirstNtupleId(1);       
  analysisManager->CreateNtuple("1", "Tuple 1");
  
  std::ostringstream os0;
  os0 <<"Dummy";
  analysisManager->CreateNtupleDColumn(os0.str());
  
  std::ostringstream os1;
  os1 <<"EnergyPrimaryMuonExitTarget";
  analysisManager->CreateNtupleDColumn(os1.str());
  
  std::ostringstream os2;
  os2 <<"EnergySecondaryPositronExitTarget";
  analysisManager->CreateNtupleDColumn(os2.str());
  
  std::ostringstream os3;
  os3 <<"EnergySecondaryElectronExitTarget";
  analysisManager->CreateNtupleDColumn(os3.str());
  
  std::ostringstream os4;
  os4 <<"EnergySecondaryMuonPlusExitTarget";
  analysisManager->CreateNtupleDColumn(os4.str());
  
  std::ostringstream os5;
  os5 <<"EnergySecondaryMuonMinusExitTarget";
  analysisManager->CreateNtupleDColumn(os5.str());
  
  std::ostringstream os6;
  os6 <<"AnglePrimaryMuonExitTarget";
  analysisManager->CreateNtupleDColumn(os6.str());
  
  std::ostringstream os7;
  os7 <<"AngleSecondaryPositronExitTarget";
  analysisManager->CreateNtupleDColumn(os7.str());
  
  std::ostringstream os8;
  os8 <<"AngleSecondaryElectronExitTarget";
  analysisManager->CreateNtupleDColumn(os8.str());
  
  std::ostringstream os9;
  os9 <<"AngleSecondaryMuonPlusExitTarget";
  analysisManager->CreateNtupleDColumn(os9.str());
  
  std::ostringstream os10;
  os10 <<"AngleSecondaryMuonMinusExitTarget";
  analysisManager->CreateNtupleDColumn(os10.str());
  
  std::ostringstream os11;
  os11 <<"ProcessID";
  analysisManager->CreateNtupleDColumn(os11.str());
  
  std::ostringstream os12;
  os12 <<"EnergyChargedExitTarget";
  analysisManager->CreateNtupleDColumn(os12.str());
  
  std::ostringstream os13;
  os13 <<"AngleChargedExitTarget";
  analysisManager->CreateNtupleDColumn(os13.str());
  
  std::ostringstream os14;
  os14 <<"EnergyChargedCut05GeVExitTarget";
  analysisManager->CreateNtupleDColumn(os14.str());
  
  std::ostringstream os15;
  os15 <<"AngleChargedCut05GeVExitTarget";
  analysisManager->CreateNtupleDColumn(os15.str());
  
  std::ostringstream os16;
  os16 <<"EnergyChargedCut1GeVExitTarget";
  analysisManager->CreateNtupleDColumn(os16.str());
  
  std::ostringstream os17;
  os17 <<"AngleChargedCut1GeVExitTarget";
  analysisManager->CreateNtupleDColumn(os17.str());
  
  std::ostringstream os18;
  os18 <<"NbTracks";
  analysisManager->CreateNtupleDColumn(os18.str());
  
  std::ostringstream os19;
  os19 <<"NbTracksCut05GeV";
  analysisManager->CreateNtupleDColumn(os19.str());
  
  std::ostringstream os20;
  os20 <<"NbTracksCut1GeV";
  analysisManager->CreateNtupleDColumn(os20.str());
  
  std::ostringstream os21;
  os21 <<"EnergyIoni";
  analysisManager->CreateNtupleDColumn(os21.str());
  
  std::ostringstream os22;
  os22 <<"EnergyPP";
  analysisManager->CreateNtupleDColumn(os22.str());
  
  std::ostringstream os23;
  os23 <<"EnergyIoniCut05GeV";
  analysisManager->CreateNtupleDColumn(os23.str());
  
  std::ostringstream os24;
  os24 <<"EnergyPPCut05GeV";
  analysisManager->CreateNtupleDColumn(os24.str());
  
  std::ostringstream os25;
  os25 <<"EnergyIoniCut1GeV";
  analysisManager->CreateNtupleDColumn(os25.str());
  
  std::ostringstream os26;
  os26 <<"EnergyPPCut1GeV";
  analysisManager->CreateNtupleDColumn(os26.str());
  
  std::ostringstream os27;
  os27 <<"AngleIoni";
  analysisManager->CreateNtupleDColumn(os27.str());
  
  std::ostringstream os28;
  os28 <<"AnglePP";
  analysisManager->CreateNtupleDColumn(os28.str());
  
  std::ostringstream os29;
  os29 <<"AngleIoniCut05GeV";
  analysisManager->CreateNtupleDColumn(os29.str());
  
  std::ostringstream os30;
  os30 <<"AnglePPCut05GeV";
  analysisManager->CreateNtupleDColumn(os30.str());
  
  std::ostringstream os31;
  os31 <<"AngleIoniCut1GeV";
  analysisManager->CreateNtupleDColumn(os31.str());
  
  std::ostringstream os32;
  os32 <<"AnglePPCut1GeV";
  analysisManager->CreateNtupleDColumn(os32.str());
  
  analysisManager->FinishNtuple(); */
  
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
